Create sampling templates when creating sampling store

JaeguKim opened the pull request:
## Which problem is this PR solving?
- Resolves #5333

## Description of the changes
- Create sampling templates when creating sampling store since doing so is intended.

## How was this change tested?
- CI

## Checklist
- [x] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [ ] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `yarn lint` and `yarn test`

yurishkuro commented on plugin/storage/es/factory.go:
```diff
@@ -317,6 +317,30 @@ func (f *Factory) CreateSamplingStore(maxBuckets int) (samplingstore.Store, erro
 	})
+
+	if f.primaryConfig.CreateIndexTemplates && !f.primaryConfig.UseILM {
+		mappingBuilder := mappings.MappingBuilder{
```
this is largely identical to L269, please refactor into a helper function that accepts a `cfg`

JaeguKim replied:
I added following helper function. https://github.com/JaeguKim/jaeger/blob/aae3616be29bcb6d2e1f7521df5e545402078825/plugin/storage/es/factory.go#L328-L340

yurishkuro reviewed (commented):
Thanks!

The full scope of this ticket includes:
  * adding tests to validate this change
  * changing existing integration test that manually creates the mapping to not do that and instead rely on this new built-in behavior

Question: what happens if the mapping already exists?

yurishkuro commented on plugin/storage/es/factory.go:
```diff
@@ -171,6 +173,8 @@ func (f *Factory) Initialize(metricsFactory metrics.Factory, logger *zap.Logger)
 		}
 	}
 
+	f.templateBuilder = es.TextTemplateBuilder{}
```
If this is a stateless class that requires no parameters to initialize, why pass it around? It can be instantiated where needed.

JaeguKim replied:
I reverted the code back to its original state. https://github.com/jaegertracing/jaeger/commit/eaa6bdce979d6f4779ffe52df358a43fcf5ac38a

JaeguKim commented:
@yurishkuro Apologies for the delayed response.
I've implemented tests to ensure coverage of code changes:
- 5c5c5b5b643f4604da8c1a005b22a0fb58edbb2d
- aae3616be29bcb6d2e1f7521df5e545402078825

I've refactored the integration test to utilize the CreateSamplingStore method instead of manual creation.
- 035603954a91113f258a0b7478a83c12fa20c1aa

When mapping was already created and SamplingStore is created again via CreateSamplingStore, this could result the existing indexes or entire indexes becoming unsearchable depending on the compatibility between existing and new index template.

I think there are several strategies.
- Merge existing index template with the new template
  - This approach involves handling cases, such as conflicting field types, to ensure that existing indexes remain searchable. 
- Create a new index template with a different name
  - By supporting search for indexes with different index template versions, we can maintain compatibility while introducing the new template. But storage cost should be considered.

yurishkuro commented:
for some reason this test consistently fails in your branch, while always succeeds in main:
```
$ go test ./plugin/storage/es
ok  	github.com/jaegertracing/jaeger/plugin/storage/es	(cached)
```

yurishkuro commented on plugin/storage/es/factory.go:
```diff
@@ -317,9 +310,35 @@ func (f *Factory) CreateSamplingStore(maxBuckets int) (samplingstore.Store, erro
 
+func getMappingBuilder(cfg *config.Configuration, templateBuilder es.TemplateBuilder) mappings.MappingBuilder {
+	return mappings.MappingBuilder{
+		TemplateBuilder:              templateBuilder,
```
why do you need this to be passed as parameter? Previously it was initialized in-place:
```
	mappingBuilder := mappings.MappingBuilder{
		TemplateBuilder:              es.TextTemplateBuilder{},
```

JaeguKim replied:
I reverted the code back to its original state. eaa6bdce979d6f4779ffe52df358a43fcf5ac38a

yurishkuro commented:
> When mapping was already created and SamplingStore is created again via CreateSamplingStore, this could result the existing indexes or entire indexes becoming unsearchable depending on the compatibility between existing and new index template.

why is that? doesn't CreateSamplingStore create **different** index templates? It should not be affecting span templates.

yurishkuro commented on plugin/storage/es/factory.go:
```diff
@@ -317,9 +310,35 @@ func (f *Factory) CreateSamplingStore(maxBuckets int) (samplingstore.Store, erro
+
+	if f.primaryConfig.CreateIndexTemplates && !f.primaryConfig.UseILM {
+		mappingBuilder := getMappingBuilder(f.primaryConfig, f.templateBuilder)
+		sampleMapping, err := mappingBuilder.GetSamplingMappings()
```
```suggestion
		samplingMapping, err := mappingBuilder.GetSamplingMappings()
```

yurishkuro commented on plugin/storage/es/factory.go:
```diff
@@ -317,9 +310,35 @@ func (f *Factory) CreateSamplingStore(maxBuckets int) (samplingstore.Store, erro
+		if err != nil {
+			return nil, err
+		}
+		if err := store.CreateTemplates(sampleMapping); err != nil {
```
can we delete CreateTemplates methods and inline it here? It's a one-liner that should not be called from anywhere else.

JaeguKim replied:
76397788b289326949557c0c7b7a988903e23fb0

JaeguKim commented:
> why is that? doesn't CreateSamplingStore create **different** index templates? It should not be affecting span templates.

https://github.com/jaegertracing/jaeger/blob/aae3616be29bcb6d2e1f7521df5e545402078825/plugin/storage/es/samplingstore/storage.go#L165

It appears that the template ID is hardcoded as jaeger-sampling. Am I interpreting this correctly?

yurishkuro commented on plugin/storage/es/factory.go:
```diff
@@ -317,9 +307,39 @@ func (f *Factory) CreateSamplingStore(maxBuckets int) (samplingstore.Store, erro
+		}
+		primaryClient := f.getPrimaryClient()
+		if primaryClient == nil {
+			return nil, errors.New("primary client is nil")
```
does this even happen? You already called getPrimaryClient in L302

yurishkuro replied:
CreateSamplingStore(), as all other CreateXYZ methods, are supposed to be called after the call to `factory.Initialize()`, which sets the primary client. So this null check is not needed.

JaeguKim replied:
0e90660db25c69d63468aafab357b974cbf92d9e

yurishkuro commented on plugin/storage/es/factory.go:
```diff
@@ -317,9 +307,39 @@ func (f *Factory) CreateSamplingStore(maxBuckets int) (samplingstore.Store, erro
+			return nil, errors.New("primary client is nil")
+		}
+		if _, err := primaryClient.CreateTemplate(indexTemplateId).Body(samplingMapping).Do(context.Background()); err != nil {
+			return nil, err
```
```suggestion
			return nil, errors.Wrap("failed to create template: %w", err)
```

yurishkuro commented on plugin/storage/es/factory.go:
```diff
@@ -317,9 +308,40 @@ func (f *Factory) CreateSamplingStore(maxBuckets int) (samplingstore.Store, erro
+		if primaryClient == nil {
+			return nil, errors.New("primary client is nil")
+		}
+		templateId := samplingTemplatePrefix + time.Now().UTC().Format(f.primaryConfig.IndexDateLayoutSampling)
```
this is not an acceptable hack. None of the other code does this, e.g. in the Writer:
```
	_, err := s.client().CreateTemplate(indexPrefix + "jaeger-span").Body(spanTemplate).Do(context.Background())
```
The template names are supposed to be stable, fixed names, not randomized strings.

JaeguKim replied:
I updated to use a fixed name and added a prefix to the sampling template name.
8118847da8567f9b95257740bb91213299022667
04ca6ff4033a8f1b2c54909c6259ff1383aeb64d

yurishkuro commented on plugin/storage/integration/elasticsearch_test.go:
```diff
@@ -100,9 +96,10 @@ func (s *ESStorageIntegration) initializeES(t *testing.T, allTagsAsFields bool)
 		DiscoverNodesOnStart: false,
 	})
 	require.NoError(t, err)
+	s.bulkProcessor, err = s.client.BulkProcessor().BulkActions(1).FlushInterval(time.Nanosecond).Do(context.Background())
```
I don't believe we need bulk processor. First, it's not actually the one that is being used to write the data (which is encapsulated in the Client). Second, I just tried an experiment (#5357) of making esRefresh function no-op, and all the tests passed just fine. I suggest removing esRefresh and removing all mentioning of bulkProcessor in this test.

JaeguKim replied:
9f801dd0d23acbc4ced4111cf9b697769838df73

yurishkuro commented on plugin/storage/integration/elasticsearch_test.go:
```diff
@@ -120,48 +117,21 @@ func (s *ESStorageIntegration) esCleanUp(t *testing.T, allTagsAsFields bool) {
 	s.logger = zaptest.NewLogger(t)
 	f := es.NewFactory()
 	v, command := config.Viperize(f.AddFlags)
+	esVersion, err := s.getVersion()
```
why is this needed? I see you pass it as an argument, but isn't it circular? When the storage is initialized it would execute the same logic to get the version: https://github.com/jaegertracing/jaeger/blob/a9a7bc22df9fd63d7eb591af2ad2c4be730b4831/pkg/es/config/config.go#L170

In fact, I would remove getVersion function from this test altogether (can be a separate PR) - the es.Client already identifies the version, we can simply expose it in its interface: https://github.com/jaegertracing/jaeger/blob/a9a7bc22df9fd63d7eb591af2ad2c4be730b4831/pkg/es/wrapper/wrapper.go#L50

JaeguKim replied:
that logic was redundant. I fixed it in commit b558e9290dd16fd79d50a1a324facab632ccf1f4.
Regarding the removal of the getVersion function, are you planning to handle it as indicated in the TODO comment in the following link?
https://github.com/jaegertracing/jaeger/blob/a9a7bc22df9fd63d7eb591af2ad2c4be730b4831/plugin/storage/integration/elasticsearch_test.go#L245

JaeguKim commented on plugin/storage/es/factory.go:
```diff
@@ -46,8 +47,9 @@ import (
-	archiveNamespace = "es-archive"
+	primaryNamespace   = "es"
+	archiveNamespace   = "es-archive"
+	samplingTemplateId = "jaeger-sampling"
```
f482946a1aba538e2c3711fa38f1bde7d956b16e

JaeguKim commented on plugin/storage/es/factory.go:
```diff
@@ -317,9 +306,39 @@ func (f *Factory) CreateSamplingStore(maxBuckets int) (samplingstore.Store, erro
+		normalizedPrefix := f.primaryConfig.IndexPrefix
+		if normalizedPrefix != "" && !strings.HasSuffix(normalizedPrefix, "-") {
+			normalizedPrefix += "-"
+		}
```
61ffa9b0b782281c1c2159914dc34db507780979

yurishkuro commented:
seeing this failure
```
    --- ❌ FAIL: TestIndexCleaner/RemoveDailyIndices_with_adaptiveSampling_prefix,_[] (2.78s)
=== RUN   TestIndexRollover_FailIfILMNotPresent
Error: ILM policy jaeger-ilm-policy doesn't exist in Elasticsearch. Please create it and re-run init
```

yurishkuro commented on plugin/storage/integration/elasticsearch_test.go:
```diff
@@ -120,48 +117,18 @@ func (s *ESStorageIntegration) esCleanUp(t *testing.T, allTagsAsFields bool) {
+		fmt.Sprintf("--es.num-shards=%v", 5),
+		fmt.Sprintf("--es.num-replicas=%v", 1),
 		fmt.Sprintf("--es.index-prefix=%v", indexPrefix),
+		fmt.Sprintf("--es.use-ilm=%v", false),
```
looks like some of the tests are dependent on ILM. Maybe you need to parameterize this.
```
    --- ❌ FAIL: TestIndexCleaner/RemoveDailyIndices_with_adaptiveSampling_prefix,_[] (2.78s)
=== RUN   TestIndexRollover_FailIfILMNotPresent
Error: ILM policy jaeger-ilm-policy doesn't exist in Elasticsearch. Please create it and re-run init
```

JaeguKim replied:
It's taking some time to pinpoint the source of the integration test failure among the recent changes I made. It appears that the ILM policy was not created in elasticsearch_test.go. I'm not sure if parameterizing it resolves the issue. I'll take a closer look at it.

yurishkuro commented:
@Pushkarm029 this is the PR I was referring to.

yurishkuro commented:
@JaeguKim could you please rebase?

JaeguKim commented:
> @JaeguKim could you please rebase?

@yurishkuro I rebased!

JaeguKim commented on plugin/storage/integration/integration.go:
```diff
@@ -299,6 +299,7 @@ func (s *StorageIntegration) testFindTraces(t *testing.T) {
 		}
 		expectedTracesPerTestCase = append(expectedTracesPerTestCase, expected)
 	}
+	time.Sleep(time.Second)
```
@yurishkuro I attempted to address the [test failure](https://github.com/jaegertracing/jaeger/actions/runs/8739681859/job/23982877531?pr=5349) occurring with ES7 by incorporating t.Parallel() to ensure the test keeps retrying until 100 seconds have elapsed. Despite this effort, the test continued to fail. While I understand that relying on time.Sleep is not ideal, I'm currently exploring alternative solutions.

yurishkuro replied:
Sleep alone is not a reliable mechanism to deal with concurrency, retries are. We already have sleep inside `waitForCondition`, which is used by `findTracesByQuery` called in this test. Why doesn't that work?

JaeguKim replied:
It turns out that the retry condition in waitForCondition was incorrect. The retry condition should match not only the trace count but also the span count. I addressed this issue in commit 8c8f2e3495b396d41a5191aae1ee3af9ad4671d6. I'm curious why the test only fails intermittently when using other Elasticsearch versions.

yurishkuro commented:
Thanks!

Pushkarm029 commented:
I think we still get this [error](https://github.com/jaegertracing/jaeger/issues/3305#issuecomment-2049269010). Because `CreateLock()` for es is yet to be implemented.

yurishkuro commented:
> Because CreateLock() for es is yet to be implemented.

yeah, we should add this to integration test
